### PR TITLE
Feature/nt 310 refactor offer decline to use exceptions

### PR DESCRIPTION
**Pull request by VarioCode**

### CODE REVIEW

File: ServiceLibrary/AlgorithmService.cs [hunk 1]
```diff
@@ -108,20 +122,11 @@ private bool TradePossible(Trade trade, Tribe originator, Tribe targetCpu)
...
-            public bool RndSelfBuild;
-            public bool RndBuildEffect;
-            public bool RndUsefulness;
-            public bool RndTradeBalance;
-
-            public int StartGoodwill;
-            public int EndGoodWill;
+            public List<OfferDeclinedException> issuesWithTrade { get; } = issues;
+            public bool tradeAccepted { get; } = tradeAccepted;
+            public Trade? coutnerOffer { get; } = counterOffer;
```
Comment (PatrickSlag) [inline ServiceLibrary/AlgorithmService.cs]:
```suggestion
            public Trade? counterOffer { get; } = counterOffer;

```

**Review by PatrickSlag — CHANGES_REQUESTED** [review]

**Review by PatrickSlag — APPROVED** [review]

File: ServiceLibrary/AlgorithmService.cs [hunk 2]
```diff
@@ -30,72 +33,83 @@ public AlgorithmService()
             _tradeBalance = new TradeBalance(random);
         }
 
-        public bool Decide(Trade trade,Tribe originator, Tribe targetCpu)
+        public void Decide(Trade trade,Tribe originator, Tribe targetCpu)
         {
-            var algoArgs = new AlgorithmDecisionEventArgs();
             int startGoodwill = targetCpu.GoodWill[originator];
+            List<OfferDeclinedException> exceptions = [];
```
Comment (PatrickSlag) [inline ServiceLibrary/AlgorithmService.cs]:
```suggestion
            List<OfferDeclinedException> exceptions = new List<OfferDeclinedException>();

```

**Review by PatrickSlag — CHANGES_REQUESTED** [review]

**Comment by PatrickSlag** [comment]
Deal needs to be reversed as in a conter offer the target becomes the originator

**Review by PatrickSlag — APPROVED** [review]
👍

**Review by jortdr — APPROVED** [review]

**Pull request merged**